### ISSUE TITLE
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed

### ISSUE THREAD

**Issue by koen84** [binance-chain/bsc-erigon#17]
#### System information

Erigon version: `./erigon --version`

OS & Version: Ubuntu 20.04 LTS

Commit hash : d9423ef

Looks like i've got several of those, all around the same blockrange.  Because of the crash, stage 6 reverts about 10k blocks
````
[INFO] [01-10|04:16:46.363] Write to db                              progress=3.9M/7.2M current table=PlainState
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|04:47:10.247] [6/16 Execution] Executed blocks         number=14083144 blk/s=11.229 tx/s=2146.761 Mgas/s=312.103 batch="508.17 MiB" alloc="1.97 GiB" sys="4.37 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|05:12:30.988] [6/16 Execution] Executed blocks         number=14083105 blk/s=10.910 tx/s=2099.872 Mgas/s=304.000 batch="507.13 MiB" alloc="1.40 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|05:37:51.905] [6/16 Execution] Executed blocks         number=14083155 blk/s=10.863 tx/s=2075.122 Mgas/s=302.699 batch="508.44 MiB" alloc="2.19 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|06:03:04.471] [6/16 Execution] Executed blocks         number=14083292 blk/s=11.089 tx/s=2090.126 Mgas/s=306.856 batch="511.99 MiB" alloc="2.46 GiB" sys="2.73 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|06:28:02.726] [6/16 Execution] Executed blocks         number=14083152 blk/s=10.782 tx/s=2061.622 Mgas/s=299.934 batch="508.34 MiB" alloc="1.67 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|06:53:13.756] [6/16 Execution] Executed blocks         number=14083270 blk/s=10.957 tx/s=2069.284 Mgas/s=302.722 batch="511.37 MiB" alloc="1.33 GiB" sys="2.71 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|07:17:45.287] [6/16 Execution] Executed blocks         number=14083239 blk/s=11.064 tx/s=2092.740 Mgas/s=305.586 batch="510.52 MiB" alloc="1.84 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|07:42:23.398] [6/16 Execution] Executed blocks         number=14082992 blk/s=10.945 tx/s=2077.769 Mgas/s=309.533 batch="504.32 MiB" alloc="1.66 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|08:07:23.860] [6/16 Execution] Executed blocks         number=14083099 blk/s=11.047 tx/s=2120.475 Mgas/s=307.499 batch="506.95 MiB" alloc="1.99 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|08:32:37.497] [6/16 Execution] Executed blocks         number=14083280 blk/s=11.045 tx/s=2079.312 Mgas/s=303.594 batch="511.61 MiB" alloc="2.06 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|08:57:21.113] [6/16 Execution] Executed blocks         number=14083180 blk/s=11.193 tx/s=2136.936 Mgas/s=311.670 batch="509.07 MiB" alloc="2.51 GiB" sys="2.71 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|09:22:47.478] [6/16 Execution] Executed blocks         number=14083204 blk/s=10.891 tx/s=2082.121 Mgas/s=304.354 batch="509.64 MiB" alloc="1.57 GiB" sys="2.73 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|09:47:55.952] [6/16 Execution] Executed blocks         number=14083266 blk/s=10.988 tx/s=2077.936 Mgas/s=302.838 batch="511.19 MiB" alloc="1.42 GiB" sys="2.72 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|10:12:36.626] [6/16 Execution] Executed blocks         number=14083120 blk/s=11.245 tx/s=2155.987 Mgas/s=311.764 batch="507.50 MiB" alloc="2.39 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|10:37:30.334] [6/16 Execution] Executed blocks         number=14083064 blk/s=10.988 tx/s=2095.965 Mgas/s=306.365 batch="506.05 MiB" alloc="2.22 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|11:02:27.755] [6/16 Execution] Executed blocks         number=14083183 blk/s=11.056 tx/s=2115.987 Mgas/s=308.006 batch="509.13 MiB" alloc="1.32 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|11:27:13.727] [6/16 Execution] Executed blocks         number=14083122 blk/s=11.231 tx/s=2154.746 Mgas/s=311.272 batch="507.56 MiB" alloc="2.04 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|11:52:25.183] [6/16 Execution] Executed blocks         number=14083191 blk/s=10.744 tx/s=2051.756 Mgas/s=299.286 batch="509.29 MiB" alloc="1.68 GiB" sys="2.72 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|12:17:03.345] [6/16 Execution] Executed blocks         number=14083180 blk/s=11.125 tx/s=2124.649 Mgas/s=309.770 batch="509.07 MiB" alloc="1.88 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|12:41:42.294] [6/16 Execution] Executed blocks         number=14083071 blk/s=11.086 tx/s=2116.001 Mgas/s=308.965 batch="506.22 MiB" alloc="1.35 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|13:06:47.661] [6/16 Execution] Executed blocks         number=14083013 blk/s=11.073 tx/s=2103.840 Mgas/s=312.281 batch="504.86 MiB" alloc="2.12 GiB" sys="2.68 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|13:32:34.045] [6/16 Execution] Executed blocks         number=14083127 blk/s=10.778 tx/s=2062.195 Mgas/s=298.412 batch="507.72 MiB" alloc="1.47 GiB" sys="2.80 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|13:57:32.939] [6/16 Execution] Executed blocks         number=14083109 blk/s=10.700 tx/s=2057.272 Mgas/s=297.264 batch="507.22 MiB" alloc="2.48 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|14:24:10.690] [p2p] GoodPeers                          eth66=29
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|14:49:24.425] [6/16 Execution] Executed blocks         number=14083217 blk/s=10.751 tx/s=2042.345 Mgas/s=297.107 batch="509.97 MiB" alloc="2.27 GiB" sys="2.79 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|15:14:29.677] [6/16 Execution] Executed blocks         number=14083218 blk/s=11.087 tx/s=2112.784 Mgas/s=308.161 batch="510.00 MiB" alloc="2.42 GiB" sys="2.71 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|15:39:16.583] [6/16 Execution] Executed blocks         number=14083160 blk/s=11.180 tx/s=2135.166 Mgas/s=311.154 batch="508.57 MiB" alloc="2.01 GiB" sys="2.75 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|16:04:48.130] [6/16 Execution] Executed blocks         number=14083008 blk/s=10.884 tx/s=2065.976 Mgas/s=306.854 batch="504.74 MiB" alloc="2.15 GiB" sys="2.68 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|16:29:52.463] [6/16 Execution] Executed blocks         number=14083143 blk/s=11.247 tx/s=2149.859 Mgas/s=312.487 batch="508.16 MiB" alloc="2.31 GiB" sys="2.72 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|16:54:36.666] [6/16 Execution] Executed blocks         number=14083012 blk/s=10.857 tx/s=2064.583 Mgas/s=306.341 batch="504.84 MiB" alloc="1.60 GiB" sys="2.68 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|17:21:23.013] [p2p] GoodPeers                          eth66=29
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|17:48:24.750] [6/16 Execution] Executed blocks         number=14083053 blk/s=10.792 tx/s=2060.754 Mgas/s=301.180 batch="505.78 MiB" alloc="2.21 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|18:15:32.429] [6/16 Execution] Executed blocks         number=14083180 blk/s=10.905 tx/s=2083.828 Mgas/s=303.819 batch="509.07 MiB" alloc="2.04 GiB" sys="2.76 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|18:40:45.210] [6/16 Execution] Executed blocks         number=14083266 blk/s=10.784 tx/s=2023.571 Mgas/s=296.360 batch="511.19 MiB" alloc="2.56 GiB" sys="2.76 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|19:06:00.431] [6/16 Execution] Executed blocks         number=14083044 blk/s=10.864 tx/s=2076.816 Mgas/s=303.808 batch="505.56 MiB" alloc="2.21 GiB" sys="2.73 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|19:31:23.260] [6/16 Execution] Executed blocks         number=14083203 blk/s=10.700 tx/s=2038.980 Mgas/s=296.875 batch="509.61 MiB" alloc="2.44 GiB" sys="2.70 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|19:56:10.324] [6/16 Execution] Executed blocks         number=14082999 blk/s=10.900 tx/s=2076.517 Mgas/s=308.729 batch="504.50 MiB" alloc="2.16 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|20:21:20.687] [6/16 Execution] Executed blocks         number=14083023 blk/s=10.954 tx/s=2092.981 Mgas/s=308.882 batch="505.12 MiB" alloc="1.72 GiB" sys="2.74 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|20:47:18.418] [6/16 Execution] Executed blocks         number=14083285 blk/s=10.732 tx/s=2017.040 Mgas/s=295.193 batch="511.75 MiB" alloc="1.39 GiB" sys="2.72 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|21:12:22.887] [6/16 Execution] Executed blocks         number=14083281 blk/s=11.057 tx/s=2082.999 Mgas/s=304.340 batch="511.65 MiB" alloc="1.74 GiB" sys="2.74 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|21:37:13.353] [6/16 Execution] Executed blocks         number=14083084 blk/s=10.554 tx/s=2019.823 Mgas/s=293.431 batch="506.55 MiB" alloc="1.66 GiB" sys="2.74 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|22:02:57.089] [6/16 Execution] Executed blocks         number=14083088 blk/s=11.121 tx/s=2125.648 Mgas/s=310.545 batch="506.65 MiB" alloc="2.14 GiB" sys="2.71 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|22:27:45.778] [6/16 Execution] Executed blocks         number=14083039 blk/s=10.863 tx/s=2076.490 Mgas/s=304.795 batch="505.46 MiB" alloc="2.09 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|22:53:06.069] [6/16 Execution] Executed blocks         number=14083181 blk/s=11.085 tx/s=2117.723 Mgas/s=308.612 batch="509.09 MiB" alloc="2.17 GiB" sys="2.69 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|23:18:12.777] [6/16 Execution] Executed blocks         number=14083275 blk/s=10.659 tx/s=1999.844 Mgas/s=293.085 batch="511.50 MiB" alloc="1.53 GiB" sys="2.73 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
--
[INFO] [01-10|23:43:13.675] [6/16 Execution] Executed blocks         number=14083140 blk/s=10.855 tx/s=2072.667 Mgas/s=301.036 batch="508.06 MiB" alloc="2.03 GiB" sys="2.67 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
````

Full crash output :
[erigon_crash_220110.log](https://github.com/binance-chain/bsc-erigon/files/7842445/erigon_crash_220110.log)

**Comment by lucas-santoni** [comment]
I experience a similar error. Full output attached at the end.

**System Information**

```
$ ./build/bin/erigon --version
erigon version 2021.11.3-alpha
```

**Commit**

```
$ git rev-parse HEAD
d9423ef0345d3a55c643add1730bbd52c50271ea
```

**Trace**

[Download here.](https://github.com/binance-chain/bsc-erigon/files/7880994/logs.txt)

**Comment by allada** [comment]
Yes I am also getting this error, but at a different stage:
```
[INFO] [01-21|03:53:40.385] [p2p] GoodPeers                          eth66=98
[INFO] [01-21|03:53:58.662] [12/16 StorageHistoryIndex] ETL [2/2] Loading into=StorageHistory current key=7aa22201... alloc="708.3 MiB" sys="4.3 GiB"
[INFO] [01-21|03:54:28.661] [12/16 StorageHistoryIndex] ETL [2/2] Loading into=StorageHistory current key=7ad7242a... alloc="691.6 MiB" sys="4.3 GiB"
[INFO] [01-21|03:54:58.662] [12/16 StorageHistoryIndex] ETL [2/2] Loading into=StorageHistory current key=7b2082af... alloc="991.3 MiB" sys="4.3 GiB"
erigon: mdbx:6368: mdbx_pnl_check: Assertion `((pl)[1]) < limit' failed.
SIGABRT: abort
PC=0xffffbbb8a060 m=26 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 78 [syscall, locked to thread]:
runtime.cgocall(0xff0ba0, 0x40540729e8)
	runtime/cgocall.go:156 +0x50 fp=0x40540729a0 sp=0x4054072960 pc=0x426540
github.com/torquem-ch/mdbx-go/mdbx._Cfunc_mdbxgo_cursor_put2(0xf6ff24000b60, 0x401a547980, 0x3c, 0x404f7ad600, 0x1e, 0x0)
	_cgo_gotypes.go:1033 +0x44 fp=0x40540729e0 sp=0x40540729a0 pc=0x8a0f84
github.com/torquem-ch/mdbx-go/mdbx.(*Cursor).Put.func1(0x404e567660, {0x401a547980, 0x3c, 0x3c}, {0x404f7ad600, 0x1e, 0x890}, 0x0)
	github.com/torquem-ch/mdbx-go@v0.22.2/mdbx/cursor.go:236 +0x98 fp=0x4054072a40 sp=0x40540729e0 pc=0x8a2ed8
github.com/torquem-ch/mdbx-go/mdbx.(*Cursor).Put(0x404e567660, {0x401a547980, 0x3c, 0x3c}, {0x404f7ad600, 0x1e, 0x890}, 0x0)
	github.com/torquem-ch/mdbx-go@v0.22.2/mdbx/cursor.go:236 +0x8c fp=0x4054072ab0 sp=0x4054072a40 pc=0x8a2ccc
github.com/ledgerwatch/erigon-lib/kv/mdbx.(*MdbxCursor).put(...)
	github.com/ledgerwatch/erigon-lib@v0.0.0-20211217093546-8d06531e4ed3/kv/mdbx/kv_mdbx.go:1029
github.com/ledgerwatch/erigon-lib/kv/mdbx.(*MdbxCursor).Put(0x4003c930e0, {0x401a547980, 0x3c, 0x3c}, {0x404f7ad600, 0x1e, 0x890})
	github.com/ledgerwatch/erigon-lib@v0.0.0-20211217093546-8d06531e4ed3/kv/mdbx/kv_mdbx.go:1317 +0xf8 fp=0x4054072b80 sp=0x4054072ab0 pc=0x8b2938
github.com/ledgerwatch/erigon-lib/etl.loadFilesIntoBucket.func1({0x401a4e9e40, 0x34, 0x34}, {0x401a547980, 0x3c, 0x3c}, {0x404f7ad600, 0x1e, 0x890})
	github.com/ledgerwatch/erigon-lib@v0.0.0-20211217093546-8d06531e4ed3/etl/collector.go:247 +0xa90 fp=0x4054072da0 sp=0x4054072b80 pc=0xb79050
github.com/ledgerwatch/erigon/eth/stagedsync.promoteHistory.func2.1({0x401a547980, 0x3c, 0x3c}, 0x401a55c730)
	github.com/ledgerwatch/erigon/eth/stagedsync/stage_indexes.go:214 +0x128 fp=0x4054072e40 sp=0x4054072da0 pc=0xfb66e8
```

**Comment by allada** [comment]
I did a little searching and found this, which references this project and likely the root cause:

https://github.com/erthink/libmdbx/issues/260

**Comment by flywukong** [comment]
@koen84   I have done a lot of tests around this problem this week. At present, the community seems to have located this problem. After I test it without a problem, https://github.com/erthink/libmdbx/issues/260 ，  I will consider mentioning this part of the repaired code to bsc-erigon later.

**Comment by erthink** [comment]
I fixed [the "root of cause"](https://github.com/erthink/libmdbx/issues/260) inside [libmdbx](https://github.com/erthink/libmdbx).
You can continue using the current version with assertions disabled (this adds speed).
But to use DB up to 8T, you will need all the fixes I have made.
Ones be merged into the libmdbx master branch soon.

**Comment by easeev** [comment]
> But to use DB up to 8T, you will need all the fixes I have made

What will happen after 8TB?

**Comment by erthink** [comment]
> What will happen after 8TB?

The 8TiB is the hard limit for DB with the default 4K sized pages, see https://github.com/erthink/libmdbx#limitations.
If you need a larger database, you should [increase the page size](https://erthink.github.io/libmdbx/group__c__settings.html#ga79065e4f3c5fb2ad37a52b59224d583e) (a power of 2, up to 64K inclusive), and will got up to 128TiB.

Your current used version (without my latest corrections), in addition to errors of excessive rigor in assertions, contains an off-by-one bug which disallowing use the last page of the 8TiB database.
I don't expect any other problems, but it's better to use a last (more tested and supported) version of the code.

**Comment by koen84** [comment]
I've had some conversations and feedback over at ergion.  It seems fixed on their `mdbx_no_assert` branch.  YMMV
Once more confirmaitons come in and that is merged into dev, this can be marked resolved.

**Comment by erthink** [comment]
> I've had some conversations and feedback over at ergion. It seems fixed on their `mdbx_no_assert` branch. YMMV Once more confirmaitons come in and that is merged into dev, this can be marked resolved.

@koen84, For clarification: 
- This issue not fixed in the `mdbx_no_assert` branch, but disabled assertion checking as a temporary workaround.
- However, the [corresponding issue](https://github.com/erthink/libmdbx/issues/260) has been [fixed](https://twitter.com/clip_dqd_kfa/status/1484532383856664576) in libmdbx, and I expect @AskAlexSharov  to update Erigon soon.

**Comment by koen84** [comment]
Thanks for clarifying.  Let me rephase : the branch helped me get passed the issue and i now have an erigon bsc node at chainhead.  Looking forward for the actual fix to be merged into dev branch.

**Comment by allada** [comment]
I can also confirm that the `mdbx_no_assert` did bypass the issue.

**Comment by erthink** [comment]
Please see the  [actual current status](https://github.com/erthink/libmdbx/issues/260#issuecomment-1020011425) and help me to fix issue completely.

**Comment by erthink** [comment]
Fixed & verified a week ago.
Please update to [libmdbx v0.11.4](https://github.com/erthink/libmdbx/releases/tag/v0.11.4).

**Comment by AskAlexSharov** [comment]
Updated
